### PR TITLE
PList, better than actual lists

### PR DESCRIPTION
**Pull request by PaulRitter**
[system]

This features the list helper type /plist/ which can basically be used like a normal list, but has way better and more logical behaviour. Also we could change its behaviour at our free will, which would be a nice add.
This is just a draft of what i thought up so far, please discuss so this will fit all of our collectively wanted behaviour of lists

:cl:
 * rscadd: added new list helper called plist

### CODE REVIEW

File: code/__HELPERS/plist.dm [hunk 1]
```diff
@@ -0,0 +1,126 @@
...
+=================
+*/
+//return at idx TODO maybe sanitize. list["[var]" fix that somehow
+plist/proc/operator[](var/idx)
+    return _list[idx]
+
+//check if list needs to be extended, so if idx is num and > then len. if idx is str we dont care
+//if idx is null, append
+plist/proc/operator[]=(idx, B)
+    _list.Insert(idx, B)
```
Comment (Exxion) [inline code/__HELPERS/plist.dm:21]: Why isn't this just `_list[idx] = B`? Inserting doesn't really make any sense here

File: code/__HELPERS/plist.dm [hunk 2]
```diff
@@ -0,0 +1,126 @@
...
+
+//actually add list as an element, not add all elements, also add everything else properly
+plist/proc/operator+(B)
+    src += B
+    return _list
```
Comment (Exxion) [inline code/__HELPERS/plist.dm]: Thanks to aliasing this is practically `+=`

**Comment by ShiftyRail** [comment]
Is the gain in ease of use worth the possible overhead from using a custom type/custom operator

If your concern is about adding a list to a list, just do it that way :
```
var/list/list_to_add = list("dab")
L += list(list_to_add)
```

File: code/__HELPERS/plist.dm (hunk 1 above)
Comment (PaulRitter) [reply code/__HELPERS/plist.dm:21]: insert features appending if idx = 0, thats why im using that here

Comment (PaulRitter) [reply code/__HELPERS/plist.dm:21]: dunno if the same thing happens when you call list[0]

File: code/__HELPERS/plist.dm (hunk 2 above)
Comment (PaulRitter) [reply code/__HELPERS/plist.dm]: oh shit thats wrong

File: code/__HELPERS/plist.dm (hunk 1 above)
Comment (Exxion) [reply code/__HELPERS/plist.dm:21]: Yes but its effect is completely different
It *inserts* an element, not changes an existing one
Also, why would you want it to append? There are already multiple other ways to do that

**Comment by PaulRitter** [comment]
imo having a costum type with costum operators just opens up our possibilities, so we can actually discuss how we want lists to behave instead of waiting what insanity lummox has in store for use next

File: code/__HELPERS/plist.dm (hunk 2 above)
Comment (PaulRitter) [reply code/__HELPERS/plist.dm]: + should only return the result, not modify the list

**Comment by Exxion** [comment]
There's practically no chance Lummox will ever change lists in any significant way

**Comment by ShiftyRail** [comment]
image
Hastily made profiling, code at 
https://gist.github.com/ShiftyRail/6995da6910f109ec2e7f2c9cb21e3255

Highlights : (it's always regular list vs plist). I'll compare TotalCPU as it includes "subprocs"

List creation : **1e-3** VS **10e-3** ; 10 times as expensive
Additions of items to lists : **1e-3** VS **10e-3** ; 10 times as expensive
Access to lists : **1e-3** VS **8e-3** ; 8 times as expensive
Adding lists to lists : **6e-3** VS **13e-3** ; 2 times as expensive

**Comment by PaulRitter** [comment]
What ShiftyRail posted basically fully convinces me. This idea sounded great in theory, but it basically died for me as soons as i found out you can't do infinite named args. Thanks for the analysis tho, maybe tell me how you did that sometime, seems like a nice tool to optimize procs & stuff.

Closing this idea, probably forever.

Ps: fuck lummox

**Comment by ShiftyRail** [comment]
It's relatively easy to do, I used your code + the `test.dm` file in the Gist, then I booted up a server, loaded the profiler and used all the verbs.

Don't click on two verbs at once, it might falsify the results because Byond takes some time to process the first one.

**Pull request closed without merging**